File: common/thalia_api.py
import os
import math
import asyncio
from .helper_functions import gather_with_concurrency

from common.bot_logger import get_logger

logger = get_logger(__name__)

THALIA_SERVER_URL = os.getenv("THALIA_SERVER_URL")
THALIA_API_URL = f"{THALIA_SERVER_URL}api/v2"


async def get_authenticated_member(client):
    response = await client.get(f"{THALIA_API_URL}/members/me/")
    return response.json()


async def get_member_by_id(client, user_id):
    response = await client.get(f"{THALIA_API_URL}/members/{user_id}/")
    return response.json()


async def get_membergroups(client):
    groups = await _get_paginated_results(client, f"{THALIA_API_URL}/activemembers/groups/")
    coros = [_get_individual_group(client, group["pk"]) for group in groups]
    return await gather_with_concurrency(2, *coros)


async def get_members(client):
    return await _get_paginated_results(client, f"{THALIA_API_URL}/members/")


async def _get_paginated_results(client, url):
    location = f"{url}?limit=100"

    response = await client.get(location)
    data = response.json()
    if "detail" in data and "throttle" in data["detail"]:
        logger.info("request throttled")
        return []
    results = data["results"]

    if data["count"] > 100:
        num = math.ceil(data["count"] / 100)

        async def get_next(url, i):
            response = await client.get(f"{url}?limit=100&offset={100*i}")
            data = response.json()
            return data["results"]

        coros = [get_next(url, i) for i in range(1, num)]
        more_results = await asyncio.gather(*coros)

        results += [item for sublist in more_results for item in sublist]

    return results


async def _get_individual_group(client, pk):
    response = await client.get(f"{THALIA_API_URL}/activemembers/groups/{pk}/")
    data = response.json()
    data["chair"] = next(filter(lambda x: x["chair"], data["members"]))["member"]
    return data

File: src/common/utils/thalia-api.ts
import Undici from 'undici';
import { HttpMethod } from 'undici/types/dispatcher';
import { Member } from '../models/member';
import { MemberGroupDetails } from '../models/memberGroup';
import { settle } from './async';

const { THALIA_SERVER_URL } = process.env;
const THALIA_API_URL = THALIA_SERVER_URL + 'api/v2';

const request = async (token: string, path: string, method: HttpMethod = 'GET') => {
  const response = await Undici.request(`${THALIA_API_URL}${path}`, {
    method,
    headers: {
      Authorization: `Bearer ${token}`
    }
  });

  if (response.statusCode > 200) {
    console.error(path, method);
    console.error(response.statusCode);
    console.error(await response.body.text());
  }

  return response;
};

export const getBackendToken = async () => {
  const { THALIA_CLIENT_ID, THALIA_CLIENT_SECRET } = process.env;
  const response = await Undici.request(`${THALIA_SERVER_URL}user/oauth/token/`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Accept: 'application/json; charset=UTF-8'
    },
    body: `grant_type=client_credentials&client_id=${THALIA_CLIENT_ID}&client_secret=${THALIA_CLIENT_SECRET}`
  });
  const result = await response.body.json();
  return result['access_token'];
};

export const getAuthenticatedMember = async (token: string) => {
  return getMember(token, 'me');
};

export const getMember = async (token: string, id: string | number): Promise<Member> => {
  const response = await request(token, `/members/${id}/`);
  return response.body.json();
};

export const getMemberGroups = async (token: string): Promise<MemberGroupDetails[]> => {
  const groups = await getPaginatedResult(token, `/activemembers/groups/`);
  return settle(groups.map((group) => getIndividualGroup(token, group['pk'])));
};

export const getMembers = async (token: string): Promise<Member[]> => {
  return getPaginatedResult(token, `/members/`);
};

const getIndividualGroup = async (token: string, id: string): Promise<MemberGroupDetails> => {
  const response = await request(token, `/activemembers/groups/${id}/`);
  const data = (await response.body.json()) as Omit<MemberGroupDetails, 'chair'>;
  return {
    ...data,
    chair: (data.members || []).find((member) => member.chair)?.member
  };
};

export const getPaginatedResult = async <T>(token: string, path: string): Promise<T[]> => {
  const response = await request(token, `${path}?limit=100`);
  const data = await response.body.json();

  if ('detail' in data && 'throttle' in data.detail) {
    return [];
  }
  const results = data['results'];

  if (data.count > 100) {
    const num = Math.ceil(data.count / 100);

    const getNext = async (i) => {
      const response = await request(token, `${path}?limit=100&offset=${100 * i}`);

      const data = await response.body.json();
      return data.results;
    };

    const nexts = await settle<any>(Array.from(Array(num).keys()).map((x) => getNext(x + 1)));
    nexts.forEach((next) => results.append(next));
  }

  return results;
};
